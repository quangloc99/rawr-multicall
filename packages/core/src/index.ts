export * from './Address';
export * from './InstructionContext';
export * from './instructions';
export * from './Bytes';
export * from './Call';
export * from './buildContract';
export * from './buildRawMulticallContract';
export * from './decodeOutput';
export * from './errors';
export * from './registerPredeployContract';
export * from './Result';
