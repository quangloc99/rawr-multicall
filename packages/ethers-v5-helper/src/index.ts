export * from './types';
export * from './createEthersV5Call';
export * from './error';
