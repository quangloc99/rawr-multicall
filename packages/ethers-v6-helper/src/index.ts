export * from './types';
export * from './error';
export * from './createEthersV6Call';
