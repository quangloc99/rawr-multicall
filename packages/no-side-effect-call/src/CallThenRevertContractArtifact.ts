import * as CallThenRevertContractArtifact from '../artifacts/contracts/CallThenRevert.yul/CallThenRevert.json';
export { CallThenRevertContractArtifact };

export const CallThenRevertContractByteCode =
    CallThenRevertContractArtifact.contracts['CallThenRevert.yul'].CallThenRevert.evm.bytecode.object;
