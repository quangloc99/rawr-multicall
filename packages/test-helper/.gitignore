artifacts
ethers-v6-contracts/typechain-types