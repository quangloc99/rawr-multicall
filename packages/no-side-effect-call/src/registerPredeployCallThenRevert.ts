import { registerPredeployContract } from '@raw-multicall/core';
import { CallThenRevertContractByteCode } from './CallThenRevertContractArtifact';

export const CallThenRevertContractLabel = '@raw-multicall/no-side-effect-call/CallThenRevertContract';
export function registerPredeployCallThenRevert() {
    registerPredeployContract(CallThenRevertContractLabel, CallThenRevertContractByteCode);
}
