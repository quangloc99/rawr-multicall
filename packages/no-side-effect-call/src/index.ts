import { registerPredeployCallThenRevert } from './registerPredeployCallThenRevert';
registerPredeployCallThenRevert();

export * from './CallThenRevertContractArtifact';
export * from './registerPredeployCallThenRevert';
export * from './wrapNoSideEffectCall';
