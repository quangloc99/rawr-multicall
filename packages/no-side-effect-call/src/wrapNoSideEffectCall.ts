import { Call, labeledAddress, assert, concatBytes, numberToBytes } from '@raw-multicall/core';
import { CallThenRevertContractLabel } from './registerPredeployCallThenRevert';

export function wrapNoSideEffectCall<ReturnType, ErrorType>(
    call: Call<ReturnType, ErrorType>
): Call<ReturnType, ErrorType> {
    return {
        getContractAddress: () => labeledAddress(CallThenRevertContractLabel),
        getData: (context) => {
            const address = context.resolveAddress(call.getContractAddress(context));
            return concatBytes([
                numberToBytes(call.getGasLimit(context) ?? 2 ** 32 - 1, 4),
                address,
                call.getData(context),
            ]);
        },
        decodeOutput: (success, data) => {
            console.log(data);
            assert(!success, 'CallThenRevert contract result should not have successful status');
            const actualSuccess = data[0];
            const actualData = data.slice(1);
            assert(actualSuccess <= 1, 'Actual success must either be 0 or 1');
            return call.decodeOutput(!!actualSuccess, actualData);
        },
        getGasLimit: () => undefined,
        getValue: (context) => call.getValue(context),
        getDependentLabeledContract: (context) => {
            const addr = call.getContractAddress(context);
            const res = call.getDependentLabeledContract?.(context) ?? [];
            if (typeof addr == 'object' && addr.type == 'labeled') return [...res, addr.label];
            return res;
        },
    };
}
