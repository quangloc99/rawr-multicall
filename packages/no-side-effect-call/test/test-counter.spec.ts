import { describeForChain } from '@rawr-multicall/test-helper';
import { Counter__factory } from '@rawr-multicall/test-helper/ethers-v6-contracts/typechain-types';
import { createEthersV6Call } from '@rawr-multicall/ethers-v6-helper';
import { ethers } from 'ethers';
import { buildRawMulticallContract, decodeRawResult, decodeResult, labeledAddress, unwrap } from '@rawr-multicall/core';
import { wrapNoSideEffectCall } from '../src';

describeForChain(
    (chain: number) => `Test counter for chain ${chain}`,
    (rpcUrl) => {
        const provider = new ethers.JsonRpcProvider(rpcUrl);
        const contract = Counter__factory.connect(ethers.ZeroAddress);

        it('test counter', async () => {
            const label = 'test-counter';
            const testContractAddress = labeledAddress(label);
            const calls = [
                await createEthersV6Call(contract, 'counter', [], { withAddress: testContractAddress }),
                await createEthersV6Call(contract, 'inc', [], { withAddress: testContractAddress }),
                await createEthersV6Call(contract, 'counter', [], { withAddress: testContractAddress }),
                wrapNoSideEffectCall(
                    await createEthersV6Call(contract, 'inc', [], { withAddress: testContractAddress })
                ),
                await createEthersV6Call(contract, 'counter', [], { withAddress: testContractAddress }),
                await createEthersV6Call(contract, 'inc', [], { withAddress: testContractAddress }),
                await createEthersV6Call(contract, 'counter', [], { withAddress: testContractAddress }),
            ] as const;
            const calldata = buildRawMulticallContract(calls, {
                predeployContracts: {
                    [label]: (await new Counter__factory().getDeployTransaction()).data,
                },
            });
            expect(calldata).toMatchSnapshot('calldata');
            const rawResult = await provider.call({ data: calldata.byteCode.toString() });
            expect(rawResult).toMatchSnapshot('rawResult');
            expect(decodeRawResult(rawResult)).toMatchSnapshot('decodedRawResult');
            const result = decodeResult(calls, rawResult);
            expect(unwrap(result[0])).toBe(0n); // counter
            expect(unwrap(result[1])).toBe(1n); // inc
            expect(unwrap(result[2])).toBe(1n); // counter
            expect(unwrap(result[3])).toBe(2n); // no side effect inc
            expect(unwrap(result[4])).toBe(1n); // counter
            expect(unwrap(result[5])).toBe(2n); // inc
            expect(unwrap(result[6])).toBe(2n); // counter
        });
    }
);
