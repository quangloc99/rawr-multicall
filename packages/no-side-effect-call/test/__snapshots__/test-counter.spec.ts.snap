// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test counter for chain 1 test counter: calldata 1`] = `
{
  "byteCode": "0x610205620001d56080397fb4cc53390356cae30ea30642519a40ac37384487b38052912725d0672a13410d61019d60806000f5507f0f1d1f8fa88713d8b865aaf46ea58512a004bbdc9872c5616a4d5e1300e48cc2604461021d6000f55060ff60e06004610285600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610265600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d0160006000601c6102696000739244b3488c0d050a0e1442268a1c754fe5ef4ba05af1841b3d841b01815281013d6000823e3d01600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610265600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d016102859003610285f360806040526000805534801561001457600080fd5b50610179806100246000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c8063371303c01461003b57806361bc221a14610059575b600080fd5b610043610077565b60405161005091906100b1565b60405180910390f35b610061610092565b60405161006e91906100b1565b60405180910390f35b60008060008154610087906100fb565b919050819055905090565b60005481565b6000819050919050565b6100ab81610098565b82525050565b60006020820190506100c660008301846100a2565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b600061010682610098565b91507fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8203610138576101376100cc565b5b60018201905091905056fea26469706673582212206af0fb1b9ae27de0253a3971a736a10365dcb6920e6ea08213eee118edb3c7b864736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffd61bc221a371303c0ffffffff799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f371303c0",
  "splittedByteCodes": [
    Uint8Array:0x610205,
    Uint8Array:0x620001d5,
    Uint8Array:0x6080,
    Uint8Array:0x39,
    Uint8Array:0x7fb4cc53390356cae30ea30642519a40ac37384487b38052912725d0672a13410d,
    Uint8Array:0x61019d,
    Uint8Array:0x6080,
    Uint8Array:0x6000,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x7f0f1d1f8fa88713d8b865aaf46ea58512a004bbdc9872c5616a4d5e1300e48cc2,
    Uint8Array:0x6044,
    Uint8Array:0x61021d,
    Uint8Array:0x6000,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x60ff,
    Uint8Array:0x60e0,
    Uint8Array:0x6004,
    Uint8Array:0x610285,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610265,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x601c,
    Uint8Array:0x610269,
    Uint8Array:0x6000,
    Uint8Array:0x739244b3488c0d050a0e1442268a1c754fe5ef4ba0,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610265,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x610285,
    Uint8Array:0x90,
    Uint8Array:0x03,
    Uint8Array:0x610285,
    Uint8Array:0xf3,
    Uint8Array:0x,
    Uint8Array:0x60806040526000805534801561001457600080fd5b50610179806100246000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c8063371303c01461003b57806361bc221a14610059575b600080fd5b610043610077565b60405161005091906100b1565b60405180910390f35b610061610092565b60405161006e91906100b1565b60405180910390f35b60008060008154610087906100fb565b919050819055905090565b60005481565b6000819050919050565b6100ab81610098565b82525050565b60006020820190506100c660008301846100a2565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b600061010682610098565b91507fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8203610138576101376100cc565b5b60018201905091905056fea26469706673582212206af0fb1b9ae27de0253a3971a736a10365dcb6920e6ea08213eee118edb3c7b864736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffd,
    Uint8Array:0x61bc221a371303c0ffffffff799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f371303c0,
  ],
  "totalValue": 0,
}
`;

exports[`Test counter for chain 1 test counter: decodedRawResult 1`] = `
[
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000000,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000001,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000001,
    "success": true,
  },
  {
    "data": Uint8Array:0x010000000000000000000000000000000000000000000000000000000000000002,
    "success": false,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000001,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000002,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000002,
    "success": true,
  },
]
`;

exports[`Test counter for chain 1 test counter: rawResult 1`] = `"0x80000020000000000000000000000000000000000000000000000000000000000000000080000020000000000000000000000000000000000000000000000000000000000000000180000020000000000000000000000000000000000000000000000000000000000000000100000021010000000000000000000000000000000000000000000000000000000000000002800000200000000000000000000000000000000000000000000000000000000000000001800000200000000000000000000000000000000000000000000000000000000000000002800000200000000000000000000000000000000000000000000000000000000000000002"`;

exports[`Test counter for chain 42161 test counter: calldata 1`] = `
{
  "byteCode": "0x610205620001d56080397fb4cc53390356cae30ea30642519a40ac37384487b38052912725d0672a13410d61019d60806000f5507f0f1d1f8fa88713d8b865aaf46ea58512a004bbdc9872c5616a4d5e1300e48cc2604461021d6000f55060ff60e06004610285600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610265600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d0160006000601c6102696000739244b3488c0d050a0e1442268a1c754fe5ef4ba05af1841b3d841b01815281013d6000823e3d01600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610265600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d01600060006004610261600073799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f5af1841b3d841b01815281013d6000823e3d016102859003610285f360806040526000805534801561001457600080fd5b50610179806100246000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c8063371303c01461003b57806361bc221a14610059575b600080fd5b610043610077565b60405161005091906100b1565b60405180910390f35b610061610092565b60405161006e91906100b1565b60405180910390f35b60008060008154610087906100fb565b919050819055905090565b60005481565b6000819050919050565b6100ab81610098565b82525050565b60006020820190506100c660008301846100a2565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b600061010682610098565b91507fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8203610138576101376100cc565b5b60018201905091905056fea26469706673582212206af0fb1b9ae27de0253a3971a736a10365dcb6920e6ea08213eee118edb3c7b864736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffd61bc221a371303c0ffffffff799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f371303c0",
  "splittedByteCodes": [
    Uint8Array:0x610205,
    Uint8Array:0x620001d5,
    Uint8Array:0x6080,
    Uint8Array:0x39,
    Uint8Array:0x7fb4cc53390356cae30ea30642519a40ac37384487b38052912725d0672a13410d,
    Uint8Array:0x61019d,
    Uint8Array:0x6080,
    Uint8Array:0x6000,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x7f0f1d1f8fa88713d8b865aaf46ea58512a004bbdc9872c5616a4d5e1300e48cc2,
    Uint8Array:0x6044,
    Uint8Array:0x61021d,
    Uint8Array:0x6000,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x60ff,
    Uint8Array:0x60e0,
    Uint8Array:0x6004,
    Uint8Array:0x610285,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610265,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x601c,
    Uint8Array:0x610269,
    Uint8Array:0x6000,
    Uint8Array:0x739244b3488c0d050a0e1442268a1c754fe5ef4ba0,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610265,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x6004,
    Uint8Array:0x610261,
    Uint8Array:0x6000,
    Uint8Array:0x73799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x610285,
    Uint8Array:0x90,
    Uint8Array:0x03,
    Uint8Array:0x610285,
    Uint8Array:0xf3,
    Uint8Array:0x,
    Uint8Array:0x60806040526000805534801561001457600080fd5b50610179806100246000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c8063371303c01461003b57806361bc221a14610059575b600080fd5b610043610077565b60405161005091906100b1565b60405180910390f35b610061610092565b60405161006e91906100b1565b60405180910390f35b60008060008154610087906100fb565b919050819055905090565b60005481565b6000819050919050565b6100ab81610098565b82525050565b60006020820190506100c660008301846100a2565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b600061010682610098565b91507fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8203610138576101376100cc565b5b60018201905091905056fea26469706673582212206af0fb1b9ae27de0253a3971a736a10365dcb6920e6ea08213eee118edb3c7b864736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffd,
    Uint8Array:0x61bc221a371303c0ffffffff799d7d6f7366b3e0b3d13e56f9bf065bc2cca08f371303c0,
  ],
  "totalValue": 0,
}
`;

exports[`Test counter for chain 42161 test counter: decodedRawResult 1`] = `
[
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000000,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000001,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000001,
    "success": true,
  },
  {
    "data": Uint8Array:0x010000000000000000000000000000000000000000000000000000000000000002,
    "success": false,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000001,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000002,
    "success": true,
  },
  {
    "data": Uint8Array:0x0000000000000000000000000000000000000000000000000000000000000002,
    "success": true,
  },
]
`;

exports[`Test counter for chain 42161 test counter: rawResult 1`] = `"0x80000020000000000000000000000000000000000000000000000000000000000000000080000020000000000000000000000000000000000000000000000000000000000000000180000020000000000000000000000000000000000000000000000000000000000000000100000021010000000000000000000000000000000000000000000000000000000000000002800000200000000000000000000000000000000000000000000000000000000000000001800000200000000000000000000000000000000000000000000000000000000000000002800000200000000000000000000000000000000000000000000000000000000000000002"`;
